### PR TITLE
Add intergration tests

### PR DESCRIPTION
**Pull request by SpacePossum**
A 1.11 version of;
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1375
Biggest diff is the caused by the diff. ways the errormanager is connected to the fixer in the branches.

### CODE REVIEW

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+            throw new InvalidArgumentException(sprintf('Level not set in configuration "%s".', $fileName));
+        }
+
+        $fixer = new Fixer();
+        $fixer->registerBuiltInFixers();
+        $allFixers = $fixer->getFixers();
+
+        $fixers = array();
+        for ($i = count($allFixers) - 1; $i >= 0; --$i) {
+            /** @var FixerInterface $fixer */
```
Comment (GrahamCampbell) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: this shouldn't be needed

Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: it is needed.
Fixer's method `getFixers` has no PHPDoc, not does the returned property;
```php
    protected $fixers = array();
```

Comment (GrahamCampbell) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: Could you add phpdoc to that method please?

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+     */
+    public function testIntegration($testFileName, $testTitle, $fixers, $input, $expected = null)
+    {
+        $this->doTestIntegration($testFileName, $testTitle, $fixers, $input, $expected);
+    }
+
+    public function getTests()
+    {
+        $fixturesDir = realpath($this->getFixturesDir());
+        $this->assertTrue(is_dir($fixturesDir), sprintf('Given fixture dir "%s" is not a directory.', $fixturesDir));
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: We should not test `getFixturesDir`, it is only a helper inside tests. Throw some kind of LogicException instead.

**Comment by keradus** [comment]
It would be great to document `*.test` file structure

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        $tests = array();
+
+        foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($fixturesDir), \RecursiveIteratorIterator::LEAVES_ONLY) as $file) {
+            if (!preg_match('/\.test$/', $file)) {
+                continue;
+            }
+
+            $test = file_get_contents($file->getRealpath());
+            $fileName = $file->getFileName();
+            if (preg_match('/--TEST--[\r\n](.*?)\s--CONFIG--[\r\n](.*?)\s--INPUT--[\r\n](.*?[\r\n]*)(?:[\r\n]--EXPECT--\s(.*)|$)/s', $test, $match)) {
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]:
```php
if (!preg_match...) {
    throw ...;
}

$testTitle = $match[1];
```

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+                $testTitle = $match[1];
+                $config = $this->getFixersFromConfig($fileName, $match[2]);
+                $input = $match[3];
+                $expected = isset($match[4]) ? $match[4] : null;
+                $tests[] = array($fileName, $testTitle, $config, $input, $expected);
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: I guess all that `testTitle, config, input, expected` variable aren't needed.

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+            } else {
+                throw new InvalidArgumentException(sprintf('Test "%s" is not valid.', $fileName));
+            }
+        }
+
+        $this->assertNotEmpty($tests, sprintf('No tests found in fixtures dir "%s".', $fixturesDir));
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: Same here, it's not the error of code we try to test but the tests itself - that someone won't write tests well. I would throw an exception here.

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+     * @param string           $testTitle
+     * @param FixerInterface[] $fixers
+     * @param string           $input
+     * @param string|null      $expected
+     */
+    protected function doTestIntegration($testFileName, $testTitle, $fixers, $input, $expected = null)
+    {
+        $errorsManager = new ErrorsManager();
+        $fixer = new Fixer();
+        $fixer->setErrorsManager($errorsManager);
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: Why not `$errorsManager = $fixer->getErrorsManager();` ?

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        $fs = new Filesystem();
+
+        $dir = $this->getTempDir();
+        $tmpDir = new \SplFileInfo($dir);
+        if (!$tmpDir->isDir()) {
+            $fs->mkdir($tmpDir);
+        }
+
+        $this->assertTrue($tmpDir->isDir(), sprintf('Given temp dir "%s" is not a directory.', $dir));
+        $this->assertTrue($tmpDir->isWritable(), sprintf('Given temp dir "%s" is not writable.', $dir));
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: It's not a real tests

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 5 above)
Comment (GrahamCampbell) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: Yeh, exception.

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+        $tmpFile = $tmpDir->getRealPath().'/tmp.php';
+        $this->assertNotFalse(@file_put_contents($tmpFile, $input), sprintf('Failed to write to tmp. file "%s".', $tmpFile));
+
+        $changed = $fixer->fixFile(new \SplFileInfo($tmpFile), $fixers, false, true, new FileCacheManager(false, null, $fixers));
+
+        if (null !== $expected) {
+            // read back the changed file content
+            $newContent = file_get_contents($tmpFile);
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: This shouldn't be needed. One IO operation more

File: Symfony/CS/Tests/IntegrationTest.php [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Symfony\CS\Tests;
+
+/**
+ * Test that parses and runs the fixture *.test files found in
+ * /Fixtures/Integration.
+ *
+ * @author SpacePossum
+ */
+class IntegrationTest extends AbstractIntegrationTest
```
Comment (keradus) [inline Symfony/CS/Tests/IntegrationTest.php]: What are the cases for extending `AbstractIntegrationTest` ? We have only one concrete class and it looks like it do all the job we may want - find all fixtures and test them.

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+        }
+
+        $fs->remove($tmpFile);
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: There is no point in creating and removing file for each tests.
Too much IO operations.

Take a look at Linter - create one file at start, override a content when we want to test new content, and remove file at the end.

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+                }
+            }
+        }
+
+        if (null === $config['level']) {
+            throw new InvalidArgumentException(sprintf('Level not set in configuration "%s".', $fileName));
+        }
+
+        $fixer = new Fixer();
+        $fixer->registerBuiltInFixers();
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: There should be one `$fixer` for all tests. Now we register built in fixers over and over again.

File: Symfony/CS/Tests/IntegrationTest.php (hunk 9 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/IntegrationTest.php]: This is an attempted to follow the suggestion of @stof here:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1375#issuecomment-137393979
happy to revert

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 1 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: I did in the other PR, then it was requested to remove it again because it is not related to this test, ie. it needs to be done in another PR.

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 4 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: these are all used AFAIK

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 6 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: because that method does not exists in 1.x

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 2 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: check, done

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 8 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: if I do that it would be after `$fs->remove($tmpFile)`,
I guess this can be done after I refactored towards your RFC here  https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1384#discussion_r38820200

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 3 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: :+1:  done (I followed https://github.com/twigphp/Twig/blob/1.x/lib/Twig/Test/IntegrationTestCase.php#L56 a bit to much..)

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 5 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: :+1:  done

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 10 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: test updated, it now re-write the tmp file every time

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 8 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: updated

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 6 above)
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: :D 
my bad, sorry :D

File: Symfony/CS/Tests/IntegrationTest.php (hunk 9 above)
Comment (keradus) [inline Symfony/CS/Tests/IntegrationTest.php]: Extended class that only override de facto 2 properties isn't much gain.

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+            $test = file_get_contents($file->getRealpath());
+            $fileName = $file->getFileName();
+            if (!preg_match('/--TEST--[\r\n](.*?)\s--CONFIG--[\r\n](.*?)\s--INPUT--[\r\n](.*?[\r\n]*)(?:[\r\n]--EXPECT--\s(.*)|$)/s', $test, $match)) {
+                throw new InvalidArgumentException(sprintf('Test "%s" is not valid.', $fileName));
+            }
+
+            $tests[] = array($fileName, $match[1], $this->getFixersFromConfig($fileName, $match[2]), $match[3], isset($match[4]) ? $match[4] : null);
+        }
+
+        if (empty($tests)) {
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: if we return an empty array doesn't PHPUnit warn about that itself?

File: Symfony/CS/Tests/IntegrationTest.php (hunk 9 above)
Comment (GrahamCampbell) [inline Symfony/CS/Tests/IntegrationTest.php]: Yeh, I agree. No point.

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+            $labelValuePair = explode('=', $line);
+            if (2 !== count($labelValuePair)) {
+                throw new InvalidArgumentException(sprintf('Invalid configuration line "%s" in "%s".', $line, $fileName));
+            }
+
+            $label = strtolower(trim($labelValuePair[0]));
+            $value = trim($labelValuePair[1]);
+
+            switch ($label) {
+                case 'level' : {
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: I didn't saw that we use braces for cases.
It may even looking good, but we should be consistent.

File: .gitignore [hunk 14]
```diff
@@ -1,3 +1,4 @@
 composer.lock
 vendor
 phpunit.xml
+Symfony/CS/Tests/Fixtures/Integration/tmp/*
```
Comment (keradus) [inline .gitignore]: It should not be needed. We should remove file at the end of tests.

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 6 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: no prloblemos :D 
let me know if I can make this easy as merge-able as possible ;)

Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: I think we need to make this PR ready to merge first ;)
Then we will think about 2.x line.

**Comment by keradus** [comment]
Travis failed.
https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer/jobs/78998867#L197

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 6 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: agreed ;)

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 12 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: yes, check removed

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 13 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: removed :+1:   (I still like those ;)  maybe a fixer for this would be nice... hmm.)

File: .gitignore (hunk 14 above)
Comment (SpacePossum) [inline .gitignore]: revered

File: Symfony/CS/Tests/IntegrationTest.php (hunk 9 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/IntegrationTest.php]: >Extended class that only override de facto 2 properties isn't much gain.

I don't think that the amount of properties or methods you have to (or can) override says much about the gain of extending a class. 
In this case, just having to implement those two abstract methods should make it easy for others to setup a test to test their own set of test files in some other directory.

Comment (keradus) [inline Symfony/CS/Tests/IntegrationTest.php]: And how it would be easier than implementing non-abstract IntegrationTest ?

**Comment by SpacePossum** [comment]
>There should be one $fixer for all tests. Now we register built in fixers over and over again.

This and remove the temp file is kind of a problem. PHPUnit has no hooks share objects between tests, nor does it have a hook to perform the deletion of the temp when the last test case has run.
The hook that is there `tearDownAfterClass` is static so it not useful since the temp file location is fetch dynamically.

Pointers about how to approach this are welcome :)

**Comment by keradus** [comment]
Why name of file can't be static then? Same for fixers table.

File: Symfony/CS/Tests/IntegrationTest.php (hunk 9 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/IntegrationTest.php]: Not easier, the Integration test is not meant to be extended. That way it can still be changed within the PHP-CS-Fixer repo without being a BC-break, maybe I should make it final...

Comment (stof) [inline Symfony/CS/Tests/IntegrationTest.php]: @keradus AbstractIntegrationTest is meant to be part of the extension points for packages extending the tool with new fixers.
What we do in Symfony or Twig is that we put such classes in the ``Test`` namespace rather than ``Tests``: the class is part of the library API, not of the testsuite (and it is then used in the testsuite itself). This is even more important when the testsuite is in a separate folder which is not autoloadable at runtime (see Twig for instance)

Comment (keradus) [inline Symfony/CS/Tests/IntegrationTest.php]: Very well.
Then I suggest to merge it as-is, but having both class `@internal`, then, at 2.x the base class and fixture format will need to be refactred - and then we made base class `@api`.

After #1383 we should be very close to 2.x

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+        }
+
+        if (!empty($config['fixers']) || !empty($config['--fixers'])) {
+            throw new \InvalidArgumentException(sprintf('Unknown fixers in configuration "%s".', implode(',', empty($config['fixers']) ? $config['--fixers'] : $config['fixers'])));
+        }
+
+        return $fixers;
+    }
+
+    public static function tearDownAfterClass()
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: from http://symfony.com/doc/current/contributing/code/standards.html

> Declare public methods first, then protected ones and finally private ones. The exceptions to this rule are the class constructor and the setUp and tearDown methods of PHPUnit tests, which should always be the first methods to increase readability;

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+                $fs = new Filesystem();
+                $fs->mkdir($dir, 0766);
+            }
+        }
+    }
+
+    /**
+     * @dataProvider getTests
+     */
+    public function testIntegration($testFileName, $testTitle, $fixers, $input, $expected = null)
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: params should be described, shortdoc would be cool too

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+    {
+        $this->doTestIntegration($testFileName, $testTitle, $fixers, $input, $expected);
+    }
+
+    public function getTests()
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: output should be described, shortdoc would be cool too

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+
+use Symfony\Component\Filesystem\Exception\IOException;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\CS\ErrorsManager;
+use Symfony\CS\FileCacheManager;
+use Symfony\CS\Fixer;
+use Symfony\CS\FixerInterface;
+
+/**
+ * Integration test helper.
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: It is not a helper ;)
Also, long desc would be cool to descr what are integration tests.

**Comment by keradus** [comment]
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1384#issuecomment-138062303

**Comment by keradus** [comment]
It looks really great, I would be very happy to merge this PR after that few markers ;)

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\CS\ErrorsManager;
+use Symfony\CS\FileCacheManager;
+use Symfony\CS\Fixer;
+use Symfony\CS\FixerInterface;
+
+/**
+ * Integration test helper. FIXME update this "It is not a helper ;)"
+ *
+ * FIXME add long description incl. fixtures file format
```
Comment (GrahamCampbell) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: todo

File: Symfony/CS/Tests/IntegrationTest.php (hunk 9 above)
Comment (stof) [inline Symfony/CS/Tests/IntegrationTest.php]: @keradus why adding the feature in an internal way first to refactor it in 2.x then ? We should implement the feature in the expected way already (the internal implementation would of course change for 2.x, but not the user API)

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 19 above)
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: changing this is pointless, it will be done before merge ;)

File: Symfony/CS/Tests/IntegrationTest.php (hunk 9 above)
Comment (keradus) [inline Symfony/CS/Tests/IntegrationTest.php]: there will be major BC break on that feature after #1383

**Comment by SpacePossum** [comment]
while the tests passes, it states:
```
[0Ktravis_time:start:06e52a26
[0K$ phpunit --verbose --coverage-clover build/logs/clover.xml
PHP Strict standards:  Static function Symfony\CS\Tests\AbstractIntegrationTest::getFixturesDir() should not be abstract in /home/travis/build/FriendsOfPHP/PHP-CS-Fixer/Symfony/CS/Tests/AbstractIntegrationTest.php on line 116
```

is this something to worry about, I don't see why this is wrong, isn't this good use of late static binding?

**Comment by keradus** [comment]
LSB is good, but in PHP one can not create method, which is both abstract and static.

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+ * Integration test base class.
+ *
+ * This test searches for '.test' fixture files in the given directory.
+ * Each fixture file will be parsed and tested against the expected result.
+ *
+ * Fixture files have the following format:
+ *
+ * --TEST--
+ * Example test description
+ *  --CONFIG--
```
Comment (gharlan) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: one space before `--CONFIG--` should be removed

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+ * level=symfony|none|psr0|psr1|psr2|symfony
+ * fixers=fixer1,fixer2,...*
+ * --fixers=fixer3,fixer4,...**
```
Comment (gharlan) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: Why `--fixers` and not `-fixers`?
Or `fixers=fixer1,fixer2,-fixer3,-fixer4`?

Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: I would not focus on that, we will change whole that --CONFIG-- section on 2.x line in favor of RulesSet

Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: >Why --fixers and not -fixers?

I took it from the command line options in the readme ;)

>fixers=fixer1,fixer2,-fixer3,-fixer4

I like that better, but if we are going to change it anyway, I leave it as is.

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 20 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: good catch, fixed

File: Symfony/CS/Tests/IntegrationTest.php (hunk 9 above)
Comment (keradus) [inline Symfony/CS/Tests/IntegrationTest.php]: And clou is that #1383 will change user API we want to use here. No more levels etc

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 22]
```diff
@@ -0,0 +1,271 @@
...
+
+        return $tests;
+    }
+
+    /**
+     * Returns the full path to directory which contains the tests.
+     *
+     * @return string
+     */
+    abstract protected static function getFixturesDir();
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: Please, let me write a comment to remember that this method must not be both abstract and static.

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 23]
```diff
@@ -0,0 +1,271 @@
...
+        $tests = array();
+
+        foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($fixturesDir), \RecursiveIteratorIterator::LEAVES_ONLY) as $file) {
+            if (!preg_match('/\.test$/', $file)) {
+                continue;
+            }
+
+            $test = file_get_contents($file->getRealpath());
+            $fileName = $file->getFileName();
+            if (!preg_match('/--TEST--[\r\n](.*?)\s--CONFIG--[\r\n](.*?)\s--INPUT--[\r\n](.*?[\r\n]*)(?:[\r\n]--EXPECT--\s(.*)|$)/s', $test, $match)) {
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: `[\r\n]` ?
one char which may be `\n` (ok, we use that a lot!) or `\r` (we won't use single `\r` at all)

Comment (GrahamCampbell) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: yeh, we need to look for 0 or 1 \r and 1 \n

Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: I would say it another way. Expect `\n` and nothing more. Follow PSR ;P

Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: btw, I saw some sad systems in which you may use `\n\r` (not `\r\n`)

Comment (GrahamCampbell) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: hmmm, but then the tests will fail on a windows install out of the box won't they?

Comment (GrahamCampbell) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: isn't the default on windows to checkout using windows style and commit using unix style?

Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: Don't know the default. And still don't know any dev that use `\r\n` in PHP code (even on Windows) or allow git to mess up with his files and change line ending. All IDEs on Win support `\n` as line end.

Comment (GrahamCampbell) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: > And still don't know any dev that use \r\n in PHP code (even on Windows) 

Yeh, I check out unix style on windows, but I think I had to change the default to do that.

Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: >And still don't know any dev that use \r\n in PHP code (even on Windows) or allow git to mess up with his files and change line ending.

I know of at least 5 developers/maintainers that do this (near me). The thing is IDEs like PHPStorm have a flag to not show white space changes. So as long as everyone on the team switches that one on (which is promoted) and doesn't look at the real changes (or github) and uses no PRs but just pushes to master; there is no problem at all ;)
(same for file permissions which Windows tends to change by default as well..)

back on topic: I'll remove the '\r' : )

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 22 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: What would you like me to add/change?
(It seems to work just fine)

Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: It does not seems to work fine, you found it already:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1384#issuecomment-138591708

From #1382:
> I know, there is a problem to create abstract static method (only reasonable solution is to create it not abstract but throw sth like NotImplementedException.

(or make them non-static, but it is not a case here)

Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: Done, as this is indeed the only way out. Which is to bad because now we throw an exception during run time for something we could know is wrong on compile time.

I'm still puzzled about why this isn't allowed in PHP, since everything to support it seems to be there (after LSB was added).

**Comment by keradus** [comment]
:+1:

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 24]
```diff
@@ -0,0 +1,277 @@
...
+
+                case 'fixers' :
+                case '--fixers' :
+                    $fixers = explode(',', $value);
+                    foreach ($fixers as $fixer) {
+                        $config[$label][] = strtolower(trim($fixer));
+                    }
+
+                    break;
+
```
Comment (GrahamCampbell) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: i thought our cs was to not have a blank line after break?

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 25]
```diff
@@ -0,0 +1,277 @@
...
+                        throw new \InvalidArgumentException(sprintf('Unknown level "%s" set in configuration in "%s", expected any of "%s".', $value, $fileName, implode(', ', array_keys($levelMap))));
+                    }
+
+                    if (null !== $config['level']) {
+                        throw new \InvalidArgumentException(sprintf('Cannot use multiple levels in configuration in "%s".', $fileName));
+                    }
+
+                    $config['level'] = $value;
+                    break;
+
```
Comment (GrahamCampbell) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: same

File: Symfony/CS/Tests/AbstractIntegrationTest.php (hunk 24 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: but have one above it? 
I always find these rules the most confusion ones, like the `return` white line rules of SF.

Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: In other words, if I puzzle a Fixer together that removes white lines below `break`, `return` and `continue` people would be OK with it to add to the repo?

Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: +throws and configuration to override wich tokens should be and not fixed ;)

Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: very well, I write down the specs and come back later with a PR (might take some weeks ;) )

white lines are removed

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 26]
```diff
@@ -0,0 +1,274 @@
...
+                    if (null !== $config['level']) {
+                        throw new \InvalidArgumentException(sprintf('Cannot use multiple levels in configuration in "%s".', $fileName));
+                    }
+
+                    $config['level'] = $value;
+                    break;
+                case 'fixers' :
+                case '--fixers' :
+                    $fixers = explode(',', $value);
```
Comment (GrahamCampbell) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: uneeded asignment

Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: thanks, done

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 27]
```diff
@@ -0,0 +1,273 @@
...
+ * Code to fix
+ * --EXPECT--
+ * Expected code after fixing***
+ *
+ *   * Additional fixers may be omitted.
+ *  ** Black listed filters may be omitted.
+ * *** When the expected block is omitted the input is expected not to
+ *     be changed by the fixers.
+ *
+ * @author SpacePossum
```
Comment (GrahamCampbell) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: missing email address

Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: I dont have one ^-^

Comment (gharlan) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: https://patch-diff.githubusercontent.com/raw/FriendsOfPHP/PHP-CS-Fixer/pull/1384.patch
line 2 ;)

Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: ha, nice find! 
oh damn, I was thinking I could PR anonymous ;)

Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]:
```
git show --format=%aE 9bd3da4
possumfromspace@gmail.com
```

works too ;P

**Comment by keradus** [comment]
It is really a great job, thank you @SpacePossum !

**Comment by keradus** [comment]
Ref #1384

**Comment by SpacePossum** [comment]
Thank you all for reviewing and all the suggestions (and finding my email address back..;) )  :)
Nice merge, especially the one to master :+1: @keradus

**Comment by keradus** [comment]
And now let me update my PR with rules/sets and FixerFactory :(
